docs: improve x/mint specs (#1873)
celestiaorg/celestia-app: x/mint/README.md

# `x/mint`

> `x/mint` is a fork of the Cosmos SDK `x/mint` module that makes some changes to the inflation mechanism. The changes were motivated by a desire for Celestia to have a predictable [inflation schedule](#inflation-schedule). See [ADR-019](../../docs/architecture/adr-019-strict-inflation-schedule.md) for more details.

Reviewer: ```suggestion
`x/mint` is a fork of the Cosmos SDK `x/mint` module that makes some changes to the inflation mechanism. The changes were motivated by a desire for Celestia to have a pre-determined [inflation schedule](#inflation-schedule). See [ADR-019](../../docs/architecture/adr-019-strict-inflation-schedule.md) for more details.
```

> `x/mint` is a fork of the Cosmos SDK `x/mint` module that makes some changes to the inflation mechanism. The changes were motivated by a desire for Celestia to have a pre-determined [inflation schedule](#inflation-schedule). See [ADR-019](../../docs/architecture/adr-019-strict-inflation-schedule.md) for more details.

Reviewer 2: [Question] I thought we had decided to utilize a standardized template for the specifications of all the components in the app, similar to the one you created for the share specs. Has there been a change in the plan?

Reviewer 2: [Suggestion] If we were to adopt a template similar to the one in the specs repository, it would be beneficial to incorporate a brief overview of the purpose of the inflation module, providing a high-level background, within the Abstract section.

Author: Discussed over Slack and let's adopt this template https://github.com/celestiaorg/celestia-app/blob/main/specs/template.md

> `x/mint` is a fork of the Cosmos SDK `x/mint` module that makes some changes to the inflation mechanism. The changes were motivated by a desire for Celestia to have a pre-determined [inflation schedule](#inflation-schedule). See [ADR-019](../../docs/architecture/adr-019-strict-inflation-schedule.md) for more details.

Reviewer 2: A link to the `x/mint` module of Cosmos SDK would be great.

| 19   | 1.50              |
| 20   | 1.50              |

- **Year** indicates the number of years elapsed since chain genesis.
- **Inflation (%)** indicates the percentage of the total supply that will be minted in the next year.

## Terms

> - **Inflation Rate**: The percentage of the total supply that will be minted each year. The inflation rate is calculated once per year on the anniversary of chain genesis based on the number of years elapsed since genesis. The inflation rate is calculated as `InitialInflationRate * ((1 - DisinflationRate) ^ YearsSinceGenesis)`. See [./types/constants.go](./types/constants.go) for the constants used in this module.

Reviewer 2: Since the `InflationRate` variable is used later in the text, would be good to introduce it here. 
``` suggestion
- **Inflation Rate**: The percentage of the total supply that will be minted each year. The inflation rate is calculated once per year on the anniversary of chain genesis based on the number of years elapsed since genesis. The inflation rate `InflationRate` is calculated as `InitialInflationRate * ((1 - DisinflationRate) ^ YearsSinceGenesis)`. See [./types/constants.go](./types/constants.go) for the constants used in this module.
```

> - **Annual Provisions**: The total amount of tokens that will be minted each year. Annual provisions are calculated once per year on the anniversary of chain genesis based on the total supply and the inflation rate. Annual provisions are calculated as `TotalSupply * InflationRate`

Reviewer 2: [Question] Does this `TotalSupply` refer to the initial total supply? or the total supply plus all the tokens minted throughout the past years?

> - **Annual Provisions**: The total amount of tokens that will be minted each year. Annual provisions are calculated once per year on the anniversary of chain genesis based on the total supply and the inflation rate. Annual provisions are calculated as `TotalSupply * InflationRate`

Reviewer 2: ```suggestion
- **Annual Provisions**: The total amount of tokens that will be minted each year. Annual provisions are calculated once per year on the anniversary of chain genesis based on the total supply and the inflation rate. Annual provisions `AnnualProvisions` are calculated as `TotalSupply * InflationRate`
```

| 19   | 1.50              |
| 20   | 1.50              |

## State

See [./types/minter.go](./types/minter.go) for the `Minter` struct which contains all of the state of this module.

## State Transitions

The `Minter` struct is updated every block via `BeginBlocker`.

> ## Begin Block

Reviewer: ```suggestion
### Begin Block
```
I feel like this should be a sub header of state transitions (maybe also events as well)

See `BeginBlocker` in [./abci.go](./abci.go).

## Events

> An event is emitted every time a block provision is minted. See `mintBlockProvision` in [./abci.go](./abci.go).

Reviewer: So that's every height right?

Author: Yep!

## Client

### CLI

```shell
$ celestia-appd query mint annual-provisions
80235005639941.760000000000000000
```

```shell
$ celestia-appd query mint genesis-time
2023-05-09 00:56:15.59304 +0000 UTC

```

```shell
$ celestia-appd query mint inflation
0.080000000000000000
```

## Params

All params have been removed from this module.

## Future Improvements

1. <https://github.com/celestiaorg/celestia-app/issues/1755>
1. <https://github.com/celestiaorg/celestia-app/issues/1757>

> 1. <https://github.com/celestiaorg/celestia-app/issues/1758>

Reviewer: Not sure if this section is necessary. These are all minor improvements and it just increases the amount of overhead as we need to remember to remove them when we complete the issue

| 19   | 1.50              |
| 20   | 1.50              |

## State

> See [./types/minter.go](./types/minter.go) for the `Minter` struct which contains all of the state of this module. The

Reviewer 3: ```suggestion
See [./types/minter.go](./types/minter.go) for the `Minter` struct which contains all of the state of this module.
```

Or maybe unfinished sentence

## State Transitions

The `Minter` struct is updated every block via `BeginBlocker`.

## Messages

N/A

## Begin Block

See `BeginBlocker` in [./abci.go](./abci.go).

## End Block

N/A

## Hooks

N/A

## Events

An event is emitted every time a block provision is minted. See `mintBlockProvision` in [./abci.go](./abci.go).

## Client

### CLI

```shell
$ celestia-appd query mint annual-provisions
80235005639941.760000000000000000
```

```shell
$ celestia-appd query mint genesis-time
2023-05-09 00:56:15.59304 +0000 UTC

```

```shell
$ celestia-appd query mint inflation
0.080000000000000000

> ```

Author: TBH not sure this is valuable and pretty concerned it will become stale

Reviewer: I would still leave the queries here

Author: sounds good will do
